src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -2,22 +2,39 @@ package org.apache.spark.datasource
 
 import cats.Eval
 import com.codahale.metrics._
-import org.apache.spark._
-
 import java.util.concurrent.ConcurrentHashMap
+import java.util.{Timer, TimerTask}
+import org.apache.spark._
+import scala.concurrent.duration._
+import scala.language.postfixOps
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
-  val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
+  val metricsMap = new ConcurrentHashMap[String, Eval[ExpiringCounter]]
+
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
+
+    def getNumberOrEmpty(getter: TaskContext => AnyVal): String =
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
```

Comment:
it'd change metric name? Then we'll lose metrics under old names?

Reply:
It will. This will be coupled with driver changes

Reply:
Splitting the PR in 2 (expiration and per-stage breakdown) + adding feature flags for those would make sense as both changes have their own risks (perf or breaking metrics somehow and doing something to TaskContext that isn't allowed in executors + if anything depended on old metric names)

Reply:
I broke it down like that in terms of commits, for the same reason.
But I coupled them in a single PR because the number of metrics will increase significantly. So, by not having them together, a new bigger perf risk arises with the great increase of metrics.

But your point is valid. If something on the driver can address the stale metrics (what I'm looking into now), then both should be paired together within a deployment.

Reply:
What was the conclusion of doing some of the logic in the driver?